curator-recipes/src/main/java/org/apache/curator/framework/recipes/shared/SharedValue.java

```diff
@@ -196,7 +196,7 @@ public boolean trySetValue(VersionedValue<byte[]> previous, byte[] newValue) thr
     private void updateValue(int version, byte[] bytes) {
         while (true) {
             VersionedValue<byte[]> current = currentValue.get();
-            if (current.getVersion() >= version) {
+            if (current.getVersion() >= version && version != Integer.MIN_VALUE) {
```

Comment:
what happens after we reach this new condition ?
is the system stuck ? 

also, as "version" is a constant here,  could early exit ?  in the beginning of the method

Reply:
Thanks @eolivelli , we try to build tmp zookeeper version and create zonde with version Integer.MAX_VALUE force and this case can reappear when trySetCount this znode. When meet this issue, SharedCount#trySetCount(VersionedValue<java.lang.Integer>, int) will always return false, then application will be stuck.

Reply:
Addendum: I still try to dig if it meet wrong version before updateValue.

Reply:
Seems that [`VersionedValue.version`](https://github.com/apache/curator/blob/72beebc504d90338ded9a68ae0f6b79f5c8986fd/curator-recipes/src/main/java/org/apache/curator/framework/recipes/shared/VersionedValue.java#L27) are `Stat.version` [from ZooKeeper](https://github.com/apache/curator/blob/72beebc504d90338ded9a68ae0f6b79f5c8986fd/curator-recipes/src/main/java/org/apache/curator/framework/recipes/shared/SharedValue.java#L251). Then it is "known" to overflow in finite time in possible large `setData` frequency. Given `default` visibility of `VersionedValue`'s visibility, I think we probably can enhance this a bit with extra `zxid`. But the crucial problem comes from ZooKeeper, it only do atomic check against `Stat.version` which is a 32-bit integer and overflow finally for long running frequently modified data. And we will finally run into corner case [`checkAndIncVersion`](https://github.com/apache/zookeeper/blob/03a36d08e257c43e8377e5549d5524805fc6b8bb/zookeeper-server/src/main/java/org/apache/zookeeper/server/PrepRequestProcessor.java#L737) exposed, `-1` is not an atomic condition which means we could write wrong node data in case of wraparound and contention.

I really hope a `check_zxid` style `setData`. I am always fearing of `setData` with `version` from old incarnation.

There is a similar case in https://lists.apache.org/thread/4o3rl49rdj5y0134df922zgc8clyt86s, which overflow `Stat.cversion`. I believed ZooKeeper was not designed and suitable for certain usages, performance degradation is desireable in wrong usages. But these overflow more sounds like bugs.

Reply:
> I really hope a check_zxid style setData.

Totally +1. It should push zookeeper to changes, which is another thing. If need we should file another thread to discuss.

Addendum, here I didn't traverse cversion and aclversion if also have the same issue at curator side.

Reply:
Sound reasonable circumvent. But when `current.getVersion()` reach `-1`, the next `trySetCount` is indeterminate.

And absolutely, we should document a bit, either this wraparound and indeterminate in `-1` or old behavior. I even think whether an exception is more appropriate ?

Reply:
> But when current.getVersion() reach -1, the next trySetCount is indeterminate.

Sorry, I don't get this meaning, the server side indead to check if version is -1. But if both `version` and `currentVersion` are -1, it will increase as expect. Do you mean that some other logic I missed at curator side? Thanks.

Reply:
It is [checkAndIncVersion](https://github.com/apache/zookeeper/blob/03a36d08e257c43e8377e5549d5524805fc6b8bb/zookeeper-server/src/main/java/org/apache/zookeeper/server/PrepRequestProcessor.java#L737) in ZooKeeper side. When the `expectedVersion` is `-1`, it means `setData` blindly.

In Curator side, that is when `getVersionedValue` reports `-1`, the next `trySetValue` will ship `-1` as `expectedVersion` to server which is apparently not what we want.

Reply:
Yes, my bad. We both say the same code segment, but I misunderstand it. I will add some javadoc for this changes moment later. Thanks again.

Reply:
One corner case when review code back, if the version overflow at server side and it is negative value such as -100 now, then restart application and `currentValue` will be initialized at curator side, and the version number will be set to UNINITIALIZED_VERSION which is -1 now. Then SharedCount will never be updated because `current.getVersion() >= version && version != Integer.MIN_VALUE` always true now.
cc @kezhuw @eolivelli What do you think about?

Reply:
I guess `-1` was chose as `UNINITIALIZED_VERSION` because of it is minimum in absent of overflow. So, now we can have `Integer.MIN_VALUE` as `UNINITIALIZED_VERSION` 😮‍💨 .